File: svws-core/src/main/java/de/svws_nrw/core/data/schueler/SchuelerListeEintrag.java
package de.svws_nrw.core.data.schueler;

import java.util.ArrayList;
import java.util.List;

import de.svws_nrw.asd.types.Geschlecht;
import de.svws_nrw.transpiler.TranspilerDTO;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotNull;
import jakarta.xml.bind.annotation.XmlRootElement;

/**
 * Diese Klasse wird bei der Kommunikation über die Open-API-Schnittstelle verwendet.
 * Sie beschreibt eine Auswahl von Daten eines Schülereintrags aus einer Liste.
 */
@XmlRootElement
@Schema(description = "ein Eintrag eines Schülers in der Schülerliste.")
@TranspilerDTO
public class SchuelerListeEintrag {

	/** Die ID des Schülers. */
	@Schema(description = "die ID des Schülers", example = "4711")
	public long id = -1L;

	/** Der Nachname des Schülers. */
	@Schema(description = "der Nachname des Schülers", example = "Mustermann")
	public @NotNull String nachname = "";

	/** Der Vorname des Schülers. */
	@Schema(description = "der Vorname des Schülers", example = "Max")
	public @NotNull String vorname = "";

	/** Das Geschlecht des Schülers (m, w, d, x - siehe {@link Geschlecht}). */
	@Schema(description = "das Geschlecht des Schülers", example = "4711")
	public @NotNull String geschlecht = "";

	/** Das Geburtsdatum des Schülers. */
	@Schema(description = "das Geburtsdatum", example = "1911-11-11")
	public String geburtsdatum;

	/** Die ID der aktuellen Klasse des Schülers.*/
	@Schema(description = "die ID der aktuellen Klasse des Schülers", example = "47")
	public @NotNull Long idKlasse = -1L;

	/** Die ID des aktuellen Jahrgangs des Schülers.*/
	@Schema(description = "die ID des aktuellen Jahrgangs des Schülers", example = "32")
	public @NotNull Long idJahrgang = -1L;

	/** Der aktuelle Jahrgang des Schülers.*/
	@Schema(description = "der aktuelle Jahrgang des Schülers", example = "09")
	public @NotNull String jahrgang = "";

	/** Die bisherige Anzahl der Jahre in der Schuleingangssphase.*/
	@Schema(description = "die bisherige Anzahl der Jahre in der Schuleingangssphase", example = "2")
	public Integer epJahre = null;

	/** Der Abiturjahrgang, falls es sich um eine Schule mit Gymnasialer Oberstufe handelt. */
	@Schema(description = "der Abiturjahrgang, falls es sich um eine Schule mit Gymnasialer Oberstufe handelt", example = "2030")
	public Integer abiturjahrgang = null;

	/** Das Kürzel der aktuellen Schulgliederung des Schülers */
	@Schema(description = "das Kürzel der aktuellen Schulgliederung des Schülers", example = "GY9")
	public @NotNull String schulgliederung = "";

	/** Die Bezeichnung des Status des Schülers (Aktiv, Extern, etc.).*/
	@Schema(description = "die Bezeichnung des Status des Schülers (Aktiv, Extern, etc.)", example = "2")
	public int status;

	/** Gibt an, ob es sich bei dem Schülerdatensatz um ein Duplikat handelt oder nicht. */
	@Schema(description = "gibt an, ob es sich bei dem Schülerdatensatz um ein Duplikat handelt oder nicht", example = "true")
	public boolean istDuplikat;

	/** Das Schulnummer bei einem externen Schüler oder null, wenn der Schüler kein externer Schüler ist. */
	@Schema(description = "die Schulnummer eines externen Schülers oder null", example = "null")
	public String externeSchulNr;

	/** Die ID des Schuljahresabschnittes des Eintrags. */
	@Schema(description = "die ID des Schuljahresabschnittes des Eintrags", example = "14")
	public @NotNull Long idSchuljahresabschnitt = -1L;

	/** Die ID des Schuljahresabschnittes des Schülers. */
	@Schema(description = "die ID des Schuljahresabschnittes des Schülers", example = "14")
	public @NotNull Long idSchuljahresabschnittSchueler = -1L;

	/** Die Liste der IDs der belegten Kurse im aktuellen Abschnit  */
	@ArraySchema(schema = @Schema(implementation = Long.class))
	public final @NotNull List<Long> kurse = new ArrayList<>();

	/**
	 * Leerer Standardkonstruktor.
	 */
	public SchuelerListeEintrag() {
		// leer
	}

	/**
	 * Vergleicht anhand der ID, ob das akutelle dasselbe Objekt, wie ein anderes übergebenes Objekt ist.
	 *
	 * @param another     das zu vergleichende Objekt
	 *
	 * @return true, falls die Objekte indentisch sind, sonst false
	 */
	@Override
	public boolean equals(final Object another) {
		return (another != null) && (another instanceof SchuelerListeEintrag) && (this.id == ((SchuelerListeEintrag) another).id);
	}

	/**
	 * Erzeugt den Hashcode zu Objekt auf Basis der ID.
	 *
	 * @return den HashCode
	 */
	@Override
	public int hashCode() {
		return Long.hashCode(id);
	}

}

File: svws-webclient/src/core/ts/src/core/data/schueler/SchuelerListeEintrag.ts
import { JavaObject } from '../../../java/lang/JavaObject';
import { ArrayList } from '../../../java/util/ArrayList';
import type { List } from '../../../java/util/List';

export class SchuelerListeEintrag extends JavaObject {

	/**
	 * Die ID des Schülers.
	 */
	public id : number = 0;

	/**
	 * Der Nachname des Schülers.
	 */
	public nachname : string = "";

	/**
	 * Der Vorname des Schülers.
	 */
	public vorname : string = "";

	/**
	 * Die ID der aktuellen Klasse des Schülers.
	 */
	public idKlasse : number = -1;

	/**
	 * Der aktuelle Jahrgang des Schülers.
	 */
	public jahrgang : string = "";

	/**
	 * Der Abiturjahrgang, falls es sich um eine Schule mit Gymnasialer Oberstufe handelt.
	 */
	public abiturjahrgang : number | null = null;

	/**
	 * Das Kürzel der aktuellen Schulgliederung des Schülers
	 */
	public schulgliederung : string = "";

	/**
	 * Die Bezeichnung des Status des Schülers (Aktiv, Extern, etc.).
	 */
	public status : number = 0;

	/**
	 * Die ID des Schuljahresabschnittes des Schülers.
	 */
	public idSchuljahresabschnitt : number = -1;

	/**
	 * Die Liste der IDs der belegten Kurse im aktuellen Abschnit
	 */
	public readonly kurse : List<number> = new ArrayList();


	public constructor() {
		super();
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.data.schueler.SchuelerListeEintrag'].includes(name);
	}

	public static transpilerFromJSON(json : string): SchuelerListeEintrag {
		const obj = JSON.parse(json);
		const result = new SchuelerListeEintrag();
		if (typeof obj.id === "undefined")
			 throw new Error('invalid json format, missing attribute id');
		result.id = obj.id;
		if (typeof obj.nachname === "undefined")
			 throw new Error('invalid json format, missing attribute nachname');
		result.nachname = obj.nachname;
		if (typeof obj.vorname === "undefined")
			 throw new Error('invalid json format, missing attribute vorname');
		result.vorname = obj.vorname;
		if (typeof obj.idKlasse === "undefined")
			 throw new Error('invalid json format, missing attribute idKlasse');
		result.idKlasse = obj.idKlasse;
		if (typeof obj.jahrgang === "undefined")
			 throw new Error('invalid json format, missing attribute jahrgang');
		result.jahrgang = obj.jahrgang;
		result.abiturjahrgang = typeof obj.abiturjahrgang === "undefined" ? null : obj.abiturjahrgang === null ? null : obj.abiturjahrgang;
		if (typeof obj.schulgliederung === "undefined")
			 throw new Error('invalid json format, missing attribute schulgliederung');
		result.schulgliederung = obj.schulgliederung;
		if (typeof obj.status === "undefined")
			 throw new Error('invalid json format, missing attribute status');
		result.status = obj.status;
		if (typeof obj.idSchuljahresabschnitt === "undefined")
			 throw new Error('invalid json format, missing attribute idSchuljahresabschnitt');
		result.idSchuljahresabschnitt = obj.idSchuljahresabschnitt;
		if ((obj.kurse !== undefined) && (obj.kurse !== null)) {
			for (const elem of obj.kurse) {
				result.kurse?.add(elem);
			}
		}
		return result;
	}

	public static transpilerToJSON(obj : SchuelerListeEintrag) : string {
		let result = '{';
		result += '"id" : ' + obj.id + ',';
		result += '"nachname" : ' + '"' + obj.nachname! + '"' + ',';
		result += '"vorname" : ' + '"' + obj.vorname! + '"' + ',';
		result += '"idKlasse" : ' + obj.idKlasse! + ',';
		result += '"jahrgang" : ' + '"' + obj.jahrgang! + '"' + ',';
		result += '"abiturjahrgang" : ' + ((!obj.abiturjahrgang) ? 'null' : obj.abiturjahrgang) + ',';
		result += '"schulgliederung" : ' + '"' + obj.schulgliederung! + '"' + ',';
		result += '"status" : ' + obj.status + ',';
		result += '"idSchuljahresabschnitt" : ' + obj.idSchuljahresabschnitt! + ',';
		if (!obj.kurse) {
			result += '"kurse" : []';
		} else {
			result += '"kurse" : [ ';
			for (let i = 0; i < obj.kurse.size(); i++) {
				const elem = obj.kurse.get(i);
				result += elem;
				if (i < obj.kurse.size() - 1)
					result += ',';
			}
			result += ' ]' + ',';
		}
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

	public static transpilerToJSONPatch(obj : Partial<SchuelerListeEintrag>) : string {
		let result = '{';
		if (typeof obj.id !== "undefined") {
			result += '"id" : ' + obj.id + ',';
		}
		if (typeof obj.nachname !== "undefined") {
			result += '"nachname" : ' + '"' + obj.nachname + '"' + ',';
		}
		if (typeof obj.vorname !== "undefined") {
			result += '"vorname" : ' + '"' + obj.vorname + '"' + ',';
		}
		if (typeof obj.idKlasse !== "undefined") {
			result += '"idKlasse" : ' + obj.idKlasse + ',';
		}
		if (typeof obj.jahrgang !== "undefined") {
			result += '"jahrgang" : ' + '"' + obj.jahrgang + '"' + ',';
		}
		if (typeof obj.abiturjahrgang !== "undefined") {
			result += '"abiturjahrgang" : ' + ((!obj.abiturjahrgang) ? 'null' : obj.abiturjahrgang) + ',';
		}
		if (typeof obj.schulgliederung !== "undefined") {
			result += '"schulgliederung" : ' + '"' + obj.schulgliederung + '"' + ',';
		}
		if (typeof obj.status !== "undefined") {
			result += '"status" : ' + obj.status + ',';
		}
		if (typeof obj.idSchuljahresabschnitt !== "undefined") {
			result += '"idSchuljahresabschnitt" : ' + obj.idSchuljahresabschnitt + ',';
		}
		if (typeof obj.kurse !== "undefined") {
			if (!obj.kurse) {
				result += '"kurse" : []';
			} else {
				result += '"kurse" : [ ';
				for (let i = 0; i < obj.kurse.size(); i++) {
					const elem = obj.kurse.get(i);
					result += elem;
					if (i < obj.kurse.size() - 1)
						result += ',';
				}
				result += ' ]' + ',';
			}
		}
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

}

export function cast_de_svws_nrw_core_data_schueler_SchuelerListeEintrag(obj : unknown) : SchuelerListeEintrag {
	return obj as SchuelerListeEintrag;
}
